table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java

```diff
@@ -148,63 +149,82 @@ public String getName() {
-          if (!mConfiguration.getBoolean(Property.ALLOW_DIFF_PART_LOC_PREFIX)
-              && !ufsUri.isAncestorOf(partitionUri)) {
-            continue;
-          }
-          hiveUfsUri = part.getSd().getLocation();
-          String partName = part.getValues().toString();
-          try {
-            partName = Warehouse.makePartName(table.getPartitionKeys(), part.getValues());
-          } catch (MetaException e) {
-            LOG.warn("Error making partition name for table {}, partition {}", tableName,
-                part.getValues().toString());
-          }
-          alluxioUri = new AlluxioURI(PathUtils.concatPath(
-              mUdbContext.getTableLocation(tableName).getPath(), partName));
-
-          // mount partition path if it is not already mounted as part of the table path mount
+        if (part.getSd() == null || part.getSd().getLocation() == null) {
+          continue;
+        }
+        String partitionUfsPath = part.getSd().getLocation();
+        AlluxioURI partitionUfsUri = new AlluxioURI(partitionUfsPath);
+        String partName = makePartName(table, part);
+        if (tableUfsUri.isAncestorOf(partitionUfsUri)) {
+          // case 1: partition is located in the directory tree of the table
+          // action: put the partition into the group
+          colocatedPartitions.add(part);
+        } else if (mConfiguration.getBoolean(Property.ALLOW_DIFF_PART_LOC_PREFIX)) {
+          // case 2: partition is NOT located in the directory tree of the table,
+          // but config says mount it anyway
+          // action: mount it individually
+          AlluxioURI partitionAlluxioUri = new AlluxioURI(PathUtils.concatPath(
+              tableAlluxioUri.getPath(), partName));
           pathTranslator.addMapping(
               UdbUtils.mountAlluxioPath(tableName,
-                  partitionUri,
-                  alluxioUri,
+                  partitionUfsUri,
+                  partitionAlluxioUri,
                   mUdbContext,
                   mConfiguration),
-              hiveUfsUri);
+              partitionUfsPath);
         }
+        // case 3: not co-located partition and table, and config not allowing them
+        // action: ignore it
+      }
+
+      // add mappings for all colocated partitions
+      for (Partition part : colocatedPartitions) {
+        String partName = makePartName(table, part);
+        pathTranslator.addMapping(
+            PathUtils.concatPath(tableAlluxioUri.getPath(), partName),
+            part.getSd().getLocation()
+        );
       }
       return pathTranslator;
     } catch (AlluxioException e) {
-      throw new IOException(
-          "Failed to mount table location. tableName: " + tableName
-              + " hiveUfsLocation: " + hiveUfsUri
-              + " AlluxioLocation: " + alluxioUri
-              + " error: " + e.getMessage(), e);
+      throw new IOException(String.format(
+          "Failed to mount table location. " +
```

Comment:
table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java:212:46: error: '+' should be on a new line.

Reply:
Yep. Fixed in 3f089c6.